File: csharp/Core/ConnectorLogger.cs
namespace Aurelion.Connector.Core;

public sealed class ConnectorLogger
{
    private readonly ConnectorConfig _cfg;

    public ConnectorLogger(ConnectorConfig cfg) => _cfg = cfg;

    public string Emit(
        string level,
        string eventType,
        string message,
        Dictionary<string, object?> payload,
        string? correlationId,
        string? causationId = null,
        string? initiatorType = null,
        string? initiatorId = null,
        string? actorType = null,
        string? actorId = null,
        string? targetType = null,
        string? targetId = null)
    {
        var eventId = Guid.NewGuid().ToString();
        var cid = correlationId ?? Guid.NewGuid().ToString();

        var logEvent = new Dictionary<string, object?>
        {
            ["event_id"] = eventId,
            ["event_type"] = eventType,
            ["level"] = level,
            ["message"] = message,
            ["timestamp"] = DateTimeOffset.UtcNow.ToString("o"),
            ["component"] = _cfg.Component,
            ["correlation_id"] = cid,
            ["payload"] = payload,
            ["initiator_type"] = initiatorType ?? "system",
            ["initiator_id"] = initiatorId ?? "platform",
            ["actor_type"] = actorType ?? "connector",
            ["actor_id"] = actorId ?? _cfg.InstanceId,
            ["target_type"] = targetType ?? "system",
            ["target_id"] = targetId ?? _cfg.InstanceId,
        };
        if (causationId is not null) logEvent["causation_id"] = causationId;

        MqPublisher.PublishJson(_cfg, _cfg.LogsExchange, "topic",
            $"{_cfg.Component}.{level}", logEvent, cid);

        return eventId;
    }
}

File: java/src/main/java/com/aurelion/connector/core/ConnectorLogger.java
package com.aurelion.connector.core;

import java.time.Instant;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.UUID;

public final class ConnectorLogger {

    private final Config cfg;

    public ConnectorLogger(Config cfg) {
        this.cfg = cfg;
    }

    public String emit(String level, String eventType, String message,
                        Map<String, Object> payload, String correlationId) {
        return emit(level, eventType, message, payload, correlationId,
                null, null, null, null, null, null, null);
    }

    public String emit(String level, String eventType, String message,
                        Map<String, Object> payload,
                        String correlationId,
                        String causationId,
                        String initiatorType, String initiatorId,
                        String actorType, String actorId,
                        String targetType, String targetId) {
        String eventId = UUID.randomUUID().toString();
        String cid = correlationId != null ? correlationId : UUID.randomUUID().toString();

        Map<String, Object> event = new LinkedHashMap<>();
        event.put("event_id", eventId);
        event.put("event_type", eventType);
        event.put("level", level);
        event.put("message", message);
        event.put("timestamp", Instant.now().toString());
        event.put("component", cfg.component());
        event.put("correlation_id", cid);
        event.put("payload", payload);
        event.put("initiator_type", initiatorType != null ? initiatorType : "system");
        event.put("initiator_id", initiatorId != null ? initiatorId : "platform");
        event.put("actor_type", actorType != null ? actorType : "connector");
        event.put("actor_id", actorId != null ? actorId : cfg.instanceId());
        event.put("target_type", targetType != null ? targetType : "system");
        event.put("target_id", targetId != null ? targetId : cfg.instanceId());
        if (causationId != null) event.put("causation_id", causationId);

        MqPublisher.publishJson(cfg, cfg.logsExchange(), "topic",
                cfg.component() + "." + level, event, cid);

        return eventId;
    }
}
